File: leetcode/queue_stack/255_Verify_Preorder_Sequence_in_Binary_Search_Tree/codes/stack.cpp
class Solution {
public:
    bool verifyPreorder(vector<int>& preorder) {
        int low = INT_MIN;
        int i = -1;
        for (int val : preorder) {
            if (val < low)
                return false;

            while (i >= 0 && val > preorder[i]) {
                low = preorder[i--];
            }

            preorder[++i] = val;
        }
        return true;
    }
};

File: leetcode/queue_stack/255_Verify_Preorder_Sequence_in_Binary_Search_Tree/codes/stack.py
from typing import List

class Solution:
    def verifyPreorder(self, preorder: List[int]) -> bool:
        stack = []
        lower_bound = float("-inf")
        
        for value in preorder:
            if value < lower_bound:
                return False
            
            while stack and value > stack[-1]:
                lower_bound = stack.pop()
            
            stack.append(value)
        
        return True
